core/chains/evm/logpoller/log_poller_internal_test.go

```diff
@@ -346,32 +346,40 @@ func TestLogPoller_Replay(t *testing.T) {
 		}()
 		assert.ErrorIs(t, lp.Replay(ctx, 4), ErrReplayInProgress)
 		<-done
+		lp.replayComplete <- nil
 	})
 
+	timeout := time.After(4 * time.Second)
+	select {
+	case <-utils.WaitGroupChan(&lp.wg):
+	case <-timeout:
+		require.Fail(t, "Log poller failed to shutdown gracefully")
```

Comment:
help me understand how this fixes the flakey test problem.

is this arbitrary timeout setting ourselves up flakey CI failures when shutdown is slow? it seems like a mask of  the underlying problem

Reply:
These lines don't fix any problem, they just allow it to exit after 4s instead of 10s if there is a problem.

The main problem was that `lp.replayComplete<-` was missing.  Even though the rest of the test before had "passed", there was still an extra goroutine left behind that could never exit.

The other problem was just passing lp.ctx to mock.On().  This can result in a race condition inside testify, since it tries to use reflect on the argument.  I think we only saw it happen that one once.

Reply:
Re:
> is this arbitrary timeout setting ourselves up flakey CI failures when shutdown is slow?

If it's possible for it to take anywhere near that long to shut down, then that's definitely a production bug that we need to be alerted to.  So this is both helping to unmask the problem (in the case that we saw), as well as detecting a potential different type of bug that we've never seen.

Reply:
For comparison:  in production, if all services combined take longer than 5s to exit, then a CRITICAL error message is logged, alerting the node op that a forced shutdown is being performed which "may result in db corruption".  Most of the top level services shut down in series rather than parallel so if anything takes more than 1s that's a problem we need to know about and fix.  This gives it 4s just in case.